### PR TITLE
2.x Changes to MultiSourcedDataHolder implementation

### PR DESCRIPTION
**Pull request by qiangdavidliu**
- updates for the same origin:name Sources replace each other regardless of source.id
- deletes only execute when origin:name:id equals for the sources

### CODE REVIEW

File: eureka2-client/src/main/java/com/netflix/eureka2/client/interest/EurekaInterestClientImpl.java [hunk 1]
```diff
@@ -78,7 +78,7 @@ public EurekaInterestClientImpl(SourcedEurekaRegistry<InstanceInfo> registry,
                     public Observable<Long> call(InterestChannel interestChannel) {
                         if (interestChannel instanceof Sourced) {
                             Source toRetain = ((Sourced) interestChannel).getSource();
-                            return registry.evictAllExcept(toRetain);
+                            return registry.evictAllExcept(Source.matcherFor(toRetain));
```
Comment (tbak) [inline eureka2-client/src/main/java/com/netflix/eureka2/client/interest/EurekaInterestClientImpl.java]: Can we rename Matcher to SourceMatcher? This name is used by many libraries.

File: eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java [hunk 2]
```diff
@@ -227,19 +230,35 @@ public void call(Throwable throwable) {
...
                 ", id='" + id + '\'' +
                 ", snapshot=" + snapshot +
                 "} " + super.toString();
     }
 
+    /**
+     * @return true if the two sources have the same origin and name. Don't use a matcher from source
+     *         to avoid the object creation.
+     */
+    private boolean matches(Source one, Source two) {
```
Comment (tbak) [inline eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java]: Can we move this method to Source class? Might be useful in other places.

File: eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java [hunk 3]
```diff
@@ -255,4 +274,94 @@ public void shutdown() {
...
+         * - If a matching source already exist in the sourceMap, remove from the dataStore first before add (as the
+         *   curr source may not equal due to a different source id).
+         * - Otherwise, just add to the dataStore
+         * - finally, add the source and sourceKey to the sourceMap
+         */
+        public void put(Source source, InstanceInfo instanceInfo) {
+            String sourceKey = sourceKey(source);
+            Source currIfExist = sourceMap.get(sourceKey);
+            if (currIfExist != null) {
+                dataMap.remove(currIfExist);
```
Comment (tbak) [inline eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java]: An update on existing entry will push it at the back of the queue, which I think is not what we want, especially for head.
If we do not remove this entry, the order will be preserved.

File: eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java [hunk 4]
```diff
@@ -255,4 +274,94 @@ public void shutdown() {
...
+            }
+            return null;
+        }
+
+        public InstanceInfo getExact(Source source) {
+            return dataMap.get(source);
+        }
+
+        /**
+         * Matches sources on origin:name only.
```
Comment (tbak) [inline eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java]: Origin, name, id?

File: eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java [hunk 5]
```diff
@@ -255,4 +274,94 @@ public void shutdown() {
             super.shutdown();
         }
     }
+
+    /**
+     * Assume access to this is synchronized.
+     * All access to the dataStore first route through the sourceMap for the authoritative source to use
+     */
+    /* visible for testing */ static class DataStore {
+        protected final Map<String, Source> sourceMap = new HashMap<>();
```
Comment (tbak) [inline eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java]: Can we use custom comparator (by origin/name) and eliminate this data structure?

File: eureka2-client/src/main/java/com/netflix/eureka2/client/interest/EurekaInterestClientImpl.java (hunk 1 above)
Comment (qiangdavidliu) [inline eureka2-client/src/main/java/com/netflix/eureka2/client/interest/EurekaInterestClientImpl.java]: sure

File: eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java (hunk 2 above)
Comment (qiangdavidliu) [inline eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java]: This is a somewhat misleading match specific to the holder, which is why it is here as a private method. The Source class already have matcher methods that does similar with a more sensible API.

File: eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java (hunk 3 above)
Comment (qiangdavidliu) [inline eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java]: Unfortunately this is not trivial unless we override Source's equals and hashcode to ignore the id, which come with its own set of issues. The whole ordering of the copies is just an optimisation anyway, so I see no strong need for it to be enforced.

File: eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java (hunk 4 above)
Comment (qiangdavidliu) [inline eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java]: updated comment

File: eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java (hunk 5 above)
Comment (qiangdavidliu) [inline eureka2-core/src/main/java/com/netflix/eureka2/registry/NotifyingInstanceInfoHolder.java]: We can use a treeMap, but we'd still have to override Source's equals and hashcode to ignore id.

**Pull request merged**